Initial work merge

mattmacf98 opened the pull request:
This is the initial work for a web based interactivity graph DCC for the new KHR_interactivity spec

Commit:
Merging Initial Work

This PR will merge initial work done on the web based DCC

Commit:
create some test graphs and clean up code

Commit:
add test graphs, update engine logic, switch rotation to quaternions

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/KhronosGroup/glTF-InteractivityGraph-AuthoringTool?pullRequest=1) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/KhronosGroup/glTF-InteractivityGraph-AuthoringTool?pullRequest=1) before we can accept your contribution.<br/><hr/>**Matthew MacFarquhar** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/KhronosGroup/glTF-InteractivityGraph-AuthoringTool?pullRequest=1) it.</sub>

Commit:
switch node ids to be implicit rather than explicit

Commit:
add special floating point nodes and comparison math nodes

Commit:
add OnSelect experimental Node, fix bug with parsing CustomEvent receive values

Commit:
fix bug in node traversal for non-unique node pointers and do sort based on numerical nodeIndex rather than lexical compare

mattmacf98 committed:
Create LICENSE

Commit:
create experimental start and stop animation nodes, fix bug with exporting to glb, update async nodes to have a done flow which executes on async completion and an out flow which executes in sync

petermart committed:
- Fixed registering all known pointers for large gltf scenes > 99 nodes
- Made sure animations and all other gltf based components are removed from cache (not just meshes) like animations.

Commit:
add authoring tool node auto alignment, update world/animateTo and world/stratAnimation and world/stopAnimation, fix bug where loaded glb KHR_interactivty graph overwrites current authored one on play

mattmacf98 committed:
Merge pull request #3 from petermart/fix-babylon-decorator

A couple basic bug fixes

Commit:
Merge branch 'initial-work-merge' of https://github.com/KhronosGroup/glTF-InteractivityGraph-AuthoringTool into initial-work-merge

Commit:
make world animate to cancellable

Commit:
Checkpoint for Demo of world/animateTo and world/set

Commit:
update failed out flows to be err instead

Commit:
world animate cubic bezier

Commit:
update forLoop and DoN to be consistent with group conversation

Commit:
add new math nodes for matrix, integers

Commit:
update throttle and delay node logic

Commit:
add isValid output value for get, refactor value evaluation logic to allow functional nodes to generate multiple output values

petermart committed:
Added some stuff from my fork.

petermart committed:
Added better auto organize. Disjoint graphs are now separated clearly.

mattmacf98 committed:
Merge pull request #5 from petermart/petermartin/better-auto-organize

Added better auto organize. Disjoint graphs are now separated clearly.

mattmacf98 committed:
Merge pull request #4 from petermart/petermartin/feburary-update

Added some inverse node, cast node, output console node, centralized fps, json parsing nan and inf, from my fork.

Commit:
sync flow nodes based on latest updates

Commit:
update throttle and delay node, fix issues with outValues  structure from other flow nodes

Commit:
update pointer paths to be pre-fixed with /, update world nodes to be pointer nodes

Commit:
BUGFIX: sequence out flows not properly connecting to other node in flows, pointer configuration not generating proper input values

Commit:
remove delay node, create setDelay and cancelDelay with non-deletable graph delay event buffer

Commit:
add boolean logic cases to logical op nodes, implement Nan node, implemement Select node, implement conversion nodes

lexaknyazev commented on src/BasicBehaveEngine/nodes/math/bitwise/Xor.ts:
```diff
@@ -0,0 +1,36 @@
+
+        let val: any;
+        switch (typeA) {
+            case "bool":
+                val = (JSON.parse(a) && !JSON.parse(b)) || (!JSON.parse(a) && JSON.parse(b));
+                break;
```
Boolean (conditional) XOR is not in the spec because it's rarely present in programming languages.

If we need to have it, the implementation should be just `JSON.parse(a) != JSON.parse(b)`.

mattmacf98 replied:
I would like to have it for completeness of full expressiveness of any business logic, (i.e. only open my door if I have clicked block 1 or block 2 but not both) I'll update the implementation in the next push

lexaknyazev replied:
OK, the spec has been updated.

Commit:
remove err flow from setDelay, fix logic for boolean XOR

Commit:
create combine and extract nodes

Commit:
create pointer interpolate node

Commit:
add pointer for khr_materials_variant

Commit:
update lifecycle, custom event and animation nodes

Commit:
change typing for float int and bool types to be singleton arrays

Commit:
add activeCamera read only pointers, make it so pointer interpolation always occurs BEFORE animation logic

Commit:
fix: make pointer interpolate  come after gltf animations

Commit:
frame graph on load, play glb asset on upload, throw unrecognized node error when loading graph

hybridherbst committed:
Fix: stopping or starting an undefined animation should not throw an exception

Commit:
add materials and material pointers

Commit:
make Custom Event and Variable config selection a drop down

Commit:
add KHR_texture_Transform pointers

mattmacf98 committed:
Merge pull request #16 from needle-tools/fix/stopping-undefined-animation

Fix: stopping or starting an undefined animation should not throw an exception

Commit:
add float2 type for all math nodes

Commit:
Merge branch 'initial-work-merge' of https://github.com/KhronosGroup/glTF-InteractivityGraph-AuthoringTool into initial-work-merge

Commit:
enable math extract nodes

Commit:
add output values from onTick, fix execution issue with how event queue is drained which caused duplicate onTick firings

Commit:
clear inlined input value when connect input value via node sockets

Commit:
properly select correct dropdown value for custom events and variables if it exists in the config

Commit:
update graph if interactivity glb is loaded

Commit:
implement KHR_selectability

Commit:
implement hoverability

hybridherbst committed:
add Node.js dependency to Prerequisites

hybridherbst committed:
add KHR_node_visibility

add partial KHR_node_visibility

full support requires a later BabylonJS version with inheritVisibility

hybridherbst committed:
fix: import and export of KHR_interactivity should not have a "graph" object

hybridherbst committed:
fix: sanitize graph layouting against routing issues in files

hybridherbst committed:
fix: adjust pointer/get and pointer/set to use "value" instead of "val"

hybridherbst committed:
feat: add "event/onSelect" log to make debugging easier

hybridherbst committed:
fix: Sequence node only worked when inputs are named like array indices, but id is purely informative

hybridherbst committed:
fix: flow/setDelay uses "done", not "complete"

mattmacf98 committed:
Merge pull request #25 from needle-tools/feature/readme-prerequisites

Add Node.js dependency to Prerequisites

mattmacf98 committed:
Merge pull request #29 from needle-tools/feature/various-fixes

Various fixes

hybridherbst committed:
rename `customEvents` -> `events`

hybridherbst committed:
properly get value in pointer/set, fix setting visibility

hybridherbst committed:
fix: Receive event values should be initialized properly

hybridherbst committed:
fix: pointer/interpolate uses "value", not "val"

hybridherbst committed:
change default engine type to Babylon

hybridherbst committed:
create default camera which does autofitting by default

robertdorn83 committed:
renamed val to value

rename keys as well, otherwise access doesn't work

robertdorn83 committed:
fixed variable set and get, was not using "value" as input/output id

robertdorn83 committed:
added more types to ADBE log

hybridherbst committed:
fix variable get/set and initial values for variables

hybridherbst committed:
add math/clamp node

mattmacf98 committed:
Merge pull request #30 from needle-tools/fix/customEvents

Fix: Rename `customEvents` -> `events` and fix pointer/set for KHR_node_visibility

mattmacf98 committed:
Merge pull request #31 from needle-tools/features/camera-fitting

Create default camera which does autofitting by default

Commit:
implement some read-only pointers

mattmacf98 committed:
Merge pull request #34 from needle-tools/features/get-set-value

Features/get set value

Commit:
add float4 to math nodes and fix some broken unit tests

Commit:
Merge branch 'initial-work-merge' of https://github.com/KhronosGroup/glTF-InteractivityGraph-AuthoringTool into initial-work-merge

hybridherbst committed:
feat: add button that can output a list of types used in this graph

robertdorn83 committed:
fixed NaN on first process in onTick timeSinceLastTick

robertdorn83 committed:
added math/mix

robertdorn83 committed:
changes singleton onStart and onTick to multiple

robertdorn83 committed:
optimized onTick with cached typeIndex

robertdorn83 committed:
changed onTick time from ms to s

robertdorn83 committed:
renamed val to value in tests

robertdorn83 committed:
added missing saturate node registry implementation

robertdorn83 committed:
changed fps from 10 to 60

robertdorn83 committed:
fixed select event position

robertdorn83 committed:
fixed warning let to const in bitwise nodes

robertdorn83 committed:
added missing bitwise node registrations

Commit:
add scrollability for variables and custom events; add not about variable initial state; and homepage to try to get live url working

mattmacf98 committed:
Add live link to README.md

robertdorn83 committed:
fixed missing math/length registration

robertdorn83 committed:
added missing math/length registrations

hybridherbst committed:
fix: register `alertParentOnHoverIn` and `alertParentOnHoverOut`

hybridherbst committed:
fix wrong array access for matrices

hybridherbst committed:
wip read-only access to `nodes/{index}/mesh` and read-write access to `mesh/{index}/primitive/{primIndex}/material`

hybridherbst committed:
wip

robertdorn83 committed:
fixed Mix node with float2, float3, float4 as C input

robertdorn83 committed:
changed activeCamera.rotationQuaternion to activeCamera.absoluteRotation

robertdorn83 committed:
fixed wrong quaternion components order in get/set rotation pointer

hybridherbst committed:
fix: cconfiguration array parsing was assuming a string, but when loading a file with flow/switch it's already an array

hybridherbst committed:
fix: register math/fract

hybridherbst committed:
register missing types

hybridherbst committed:
fix: cached values should also run through parseType

hybridherbst committed:
dev: log error when a single-entry array is encountered after evaluation

robertdorn83 committed:
set correct absRotation for active camera to xyzw

robertdorn83 committed:
revert OnTick.TimeSinceLastTick to NaN on first tick

robertdorn83 committed:
renamed onTickStartIndices to onStartIndices

robertdorn83 committed:
removed unnessary -1 check

robertdorn83 committed:
added note to readme for non normative writable pointer path for mesh material

robertdorn83 committed:
addedcomment for LHS vs RHS conversion between Babylon and Gltf

robertdorn83 committed:
removed duplicate math/saturate node spec

robertdorn83 committed:
removed math/length node spec duplicates

Commit:
properly parse arrayish booleans (i.e. [true]) when extracting values from authoring section

mattmacf98 committed:
Merge pull request #42 from needle-tools/feature/various-fixes

Various improvements

Commit:
remove old Interpolate node (in favor of Mix which was added), add not about possible incorrect mesh mapping for interactivity world, update unit tests to bring up to date with current spec langauge

Commit:
authoring tool can create basic spec compliant graphs (no configurations)

Commit:
authoring tool is able to do simple configurations

Commit:
enable variable configuration

Commit:
extarct graph state management to a context

Commit:
enable laoding from json upload

Commit:
enable loading graph to authoring component when it is given as json

Commit:
fix issue with configurations wiping values and flows in authoring

Commit:
very basic excution working in logging engine for subset of nodes

Commit:
switch all nodes to obey current json objectish values

Commit:
core nodes specified

Commit:
update behavior of pointer, events and sequence to match spec, create variable interpolated node

Commit:
enable persistance of positioning metdata

Commit:
update authoring graph UI for pointer type config dropdowns and variable/interpolate properly grabbing value type

Commit:
update loader extension to grab graph from proper location

Commit:
fix issue where edgeDelet function was not using latest graph; fix issue where configuration update used stale config values

Commit:
update spelling for declarations

Commit:
add onSelect exetension node

Commit:
implement hoverabilityt and custom ADBE log nodes

Commit:
old node unit tests adpated and working

Commit:
load unkown nodes properly

Commit:
add some test graphs

Commit:
configure custom event listeners to only trigger during a new tick

lexaknyazev commented on src/BasicBehaveEngine/decorators/LoggingDecorator.ts:
```diff
@@ -0,0 +1,142 @@
+                const value = [linearFloat(p.y, initialValue[0], targetValue[0]), linearFloat(p.y, initialValue[1], targetValue[1]), linearFloat(p.y, initialValue[2], targetValue[2])]
+                this.behaveEngine.setPathValue(path, value);
+            } else if (valueType === "float4") {
+                const value = slerpFloat4(t, initialValue, targetValue);
```
Not all `float4` values are quaternions. For example, `pbrMetallicRoughness. baseColorFactor`.

Commit:
support float2x2 and 3x3 matrices; allow pointer interpolate to picke between slerp vs lerp for float4 pointers; add math/round

Commit:
add test for handling unknown nodes that are properly declared

lexaknyazev commented on src/BasicBehaveEngine/nodes/math/vector/VectorLength.ts:
```diff
@@ -0,0 +1,35 @@
+        const type: string = this.getType(typeIndex);
+        let val: any;
+
+        switch (type) {
+            case "float2":
+                val = Math.sqrt(Math.pow(a[0],2) + Math.pow(a[1],2));
+                break;
+            case "float3":
+                val = Math.sqrt(Math.pow(a[0],2) + Math.pow(a[1],2) + Math.pow(a[2],2));
+                break;
+            case "float4":
+                val = Math.sqrt(Math.pow(a[0],2) + Math.pow(a[1],2) + Math.pow(a[2],2) + Math.pow(a[3],2));
+                break;
```
This could be reduced to:
```js
switch (type) {
    case "float2":
    case "float3":
    case "float4":
        val = Math.hypot(...a);
        break;
```

mattmacf98 replied:
Ok, I’ll update it in my next push

Commit:
switch to use math.hypot for vector length; only wipe pointer values on type change if the underlying type actually changes (this fixes a bug with loading from json wiping the value to null every time)

Commit:
add newly add float2x2 and float3x3 types for math nodes which use floatN

Commit:
fix issue with graph auto laying not running since authroing graph data retains uuid and not numerical ids

Commit:
fix getworldmatrix pointer to output float4x4; update active camera pointers; fix issue with pointer/ wiping connected values on load; fix logic for variable get and set; fix issue where disconnected flows were not getting removed from context; fix input value names for extract nodes

Commit:
fix issue where pointer/get nodes got an input value socket on load from json

Commit:
fix camera pointers to account for LHS system; add two quat nodes for math

Commit:
move varibale and pointer interpolation to be handled as part of the tick event loop

Commit:
fix bug with pointer input values being wiped on load even though the loaded asset has values for the socket; use type signature if name property is not present for types dropdown in UI

Commit:
upgrade babylon and use babylon implemented node_visibility

lexaknyazev commented on src/BasicBehaveEngine/nodes/math/vector/Normalize.ts:
```diff
@@ -0,0 +1,53 @@
+        switch (type) {
+            case "float2":
+                // eslint-disable-next-line no-case-declarations
+                const length2 = Math.sqrt(Math.pow(a[0], 2) + Math.pow(a[1], 2));
```
Use `Math.hypot` here similarly to the `math/length` node.

mattmacf98 replied:
ok

Commit:
use math/hypot for normalize node

Commit:
add special case for socket setting for sequence since it has out flows not in spec nor determined by config

Commit:
fix math/atan2 node specs

Commit:
force all loaded nodes to inherit visibility

Commit:
when checking if node selected, use uniqueId instead of just mesh.id (which could be shared by multiple nodes)

lexaknyazev commented on src/BasicBehaveEngine/nodes/flow/MultiGate.ts:
```diff
@@ -0,0 +1,55 @@
+        this._numberOutputFlows = Object.keys(this.flows).length;
+        this._isRandom = JSON.parse(isRandom);
+        this._loop = JSON.parse(loop);
+        this._unSeenOutFlows = [...Array(this._numberOutputFlows).keys()];
```
Output flow sockets must be sorted like in `flow/sequence`.

mattmacf98 replied:
done

Commit:
do not reset cached values for pointer get node

Commit:
add clearing eval cache for all flow nodes before evaluating

Commit:
update multigate to comply with spec

Commit:
add ? for node selectability so null coallese can default it to true if not present

Commit:
for customEvent/send and variable/set add value sockets to the inputValuesToSet record so they don't get forgotten on config evaluation

Commit:
properly handle setting node's mesh prims visibility

Commit:
add animation pointer paths

hybridherbst committed:
fix showing node declaration list

Commit:
add out val index for for loop

Commit:
add proper defaulting to event defaults or type defaults for custom event receieve

Commit:
add defaulting configurations for missing config values

Commit:
strip typeOptions out of input value properties when serializing to json

Commit:
add matrix compose and decompose

lexaknyazev commented on src/BasicBehaveEngine/nodes/math/matrix/matDecompose.ts:
```diff
@@ -0,0 +1,77 @@
+            return result;
+        }
+
+        const s_x = Math.sqrt(a[0][0] * a[0][0] + a[1][0] * a[1][0] + a[2][0] * a[2][0]);
+        const s_y = Math.sqrt(a[0][1] * a[0][1] + a[1][1] * a[1][1] + a[2][1] * a[2][1]);
+        const s_z = Math.sqrt(a[0][2] * a[0][2] + a[1][2] * a[1][2] + a[2][2] * a[2][2]);
```
nit: use `Math.hypot`.

lexaknyazev commented on src/BasicBehaveEngine/nodes/math/matrix/matDecompose.ts:
```diff
@@ -0,0 +1,77 @@
+        const s_y = Math.sqrt(a[0][1] * a[0][1] + a[1][1] * a[1][1] + a[2][1] * a[2][1]);
+        const s_z = Math.sqrt(a[0][2] * a[0][2] + a[1][2] * a[1][2] + a[2][2] * a[2][2]);
+        // check scale values are non NaN and non infinite
+        if (isNaN(s_x) || isNaN(s_y) || isNaN(s_z) || !isFinite(s_x) || !isFinite(s_y) || !isFinite(s_z)) {
```
`isFinite` is false for NaN, so `isNaN` is not needed here.

lexaknyazev commented on src/BasicBehaveEngine/nodes/math/matrix/matDecompose.ts:
```diff
@@ -0,0 +1,77 @@
+            return result;
+        }
+
+        result.translation.value = [a[0][3]/ s_x, a[1][3]/ s_y, a[2][3]/ s_z];
```
Scale factors do not affect translation.

Commit:
add math/switch node and fix issue with mat compose and decompose scaling translation

Commit:
add matMul and inverse to engine, fix matrix pointer getter

Commit:
support loading other type arrays by converting inline vales, decleration vals, variables, and custom event vals to the standard type array

Commit:
use column major for all nodes that deal with matrices

Commit:
bug fixes for updating pointer type config; loading type updated graph from glb; fix matrix inverse; fix matrix pointer getter

hybridherbst committed:
remove array brackets from string

mattmacf98 committed:
Merge pull request #58 from needle-tools/fix/show-node-list

Fix leftover code from previous merge

Commit:
add small delay before creating edges to avoid react flow handle not inited issue; clean up some options to use default value insad of selected prop

Commit:
Merge branch 'initial-work-merge' of https://github.com/KhronosGroup/glTF-InteractivityGraph-AuthoringTool into initial-work-merge

Commit:
implement experimental debug node

Commit:
improve formatting of experimental debug node

Commit:
explicitly check for undefinedness (not just falsy) for dropwdown selection to avoid turning 0 value to -1

Commit:
fix case configuration parsing

Commit:
make significant progress on matrix tests (all pass expect last two from offset scale ones)

Commit:
fix bug in global matrix pointer to properly flip z scale to revert babylon inversion

Commit:
fix npm install using overrides; fix some matrix unit tests

Commit:
add cypress testing for E2E tests

Commit:
add cypress tests for matrix test graphs

Commit:
add math tests

Commit:
update hover every frame (not just when pointer is moving)

Commit:
update debug node with severity config

Commit:
update rotate2D and rotate3D nodes

Commit:
add quaternion nodes

Commit:
bug fix for eslint build error
